### ISSUE TITLE
app is very slow in android Samsung 23 ultra 

### ISSUE THREAD

**Issue by rubbishCoder** [facebook/react-native#39183]
### Description

App is very slow in Samsung devices especially in the android version 13 and also in the android logcat I get error

`
2023-08-28 17:37:37.826  1288-1288  LayerHistory | surfaceflinger | MainActivity$_27693#87452 Max (can't resolve refresh rate)
`

### React Native Version

0.72.4

### Output of `npx react-native info`

System:
  OS: macOS 13.5.1
  CPU: (10) arm64 Apple M1 Pro
  Memory: 152.75 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.5.1
    path: /opt/homebrew/bin/node
  Yarn:
    version: 1.22.19
    path: /usr/local/bin/yarn
  npm:
    version: 9.8.1
    path: /opt/homebrew/bin/npm
  Watchman:
    version: 2023.08.14.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.12.1
    path: /opt/homebrew/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 22.4
      - iOS 16.4
      - macOS 13.3
      - tvOS 16.4
      - watchOS 9.4
  Android SDK: Not Found
IDEs:
  Android Studio: 2022.3 AI-223.8836.35.2231.10671973
  Xcode:
    version: 14.3.1/14E300c
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 11.0.20
    path: /usr/bin/javac
  Ruby:
    version: 2.6.10
    path: /usr/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.72.4
    wanted: 0.72.4
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false

### Steps to reproduce

 I just installed my app and run on the Samsung s23 ultra it was running fine but very slow and very bad response when I navigate. It takes around 10 seconds just to navigate, this issue occurs only in Samsung android devices. Is there any fix to it. 
 

### Snack, screenshot, or link to a repository

 https://snack.expo.dev/qb7Cur_lU

**Comment by AndreiCalazans** [comment]
This seems unrelated to React Native core per say. What's likely impacting your navigation could be: the navigation library; a different JavaScript engine; animations; etc.,. 

Can you add more information:

1. What libraries are you using for navigation
2. What JS engine are you using
3. Is there any animations happening and if so how are they being executed
4. Can you enable [the performance monitor](https://reactnative.dev/docs/performance#what-you-need-to-know-about-frames) and see which thread is dropping frames when navigating (this can help indicate where to look)
5. Does this also happen if you use a bare react native app with react-navigation on that phone? 


And if this is a problem with core React Native a reproducible repo would help initialize an investigation. If this is core problem you should be able to create a reproduction with just a bare app and no external libraries.

**Comment by rubbishCoder** [comment]
@AndreiCalazans animations or the navigation is not a problem, I thing the axis call is the one that cause the lag.

I tried to stop all the api calls and navigated between pages it was fine and rendering fast but when I do async or synchronous  api calls the app is slow.

Using `"axios": "1.5.0"` and also I tried fetch which didn't have any impact, the app is still slow and navigations are very laggy if I make api calls, and this is occurring only in the android version 13 especially on the Samsung latest device like s22, s23. 

Using below packages for navigation.

`
    "@react-navigation/bottom-tabs": "^6.0.9",
    "@react-navigation/native": "^6.1.7",
    "@react-navigation/native-stack": "^6.2.5",
    "@react-navigation/stack": "^6.3.9",
`

Using Hermes js engine.

Lottie animation json is used to render animations ` "lottie-react-native": "^6.2.0"`.

Will create a repo with reproducible issue and update it.

**Comment by devoren** [comment]
May be related to https://github.com/facebook/react-native/issues/36296

**Comment by AndreiCalazans** [comment]
Possibly relate dto #36296, @rubbishCoder did you experience this slow down after upgrading? And yea, a reproduction sample can help us profile and see what's going on.

**Comment by rubbishCoder** [comment]
@AndreiCalazans @devoren even after updating to latest version I'm facing this issue. And I'm not able to recreate this in a different project

**Comment by devoren** [comment]
@rubbishCoder Can you upgrade to version 0.71.8 and see if there is a performance improvement?

**Comment by rubbishCoder** [comment]
> May be related to #36296

@devoren this says that it was fixed 0.72 but I'm in 0.72.4 and still facing this issue

**Comment by rubbishCoder** [comment]
> @rubbishCoder Can you downgrade to version 0.71.8 and see if there is a performance improvement?

@devoren okay I will downgrade and check it

**Comment by rubbishCoder** [comment]
@devoren no effect after downgrading to 0.71.8

**Comment by devoren** [comment]
@rubbishCoder oh, this is bad, then we just have to wait for a patch or release from the RN developers :(, in my case 0.71.8 is much better than 0.72

**Comment by rubbishCoder** [comment]
@devoren does redux have anything to do with it?

**Comment by devoren** [comment]
@rubbishCoder I think this is a problem on the engine side and RN Core. Redux is only on the js side so it doesn't affect that much, but if you have a large store and heavy functions then yes Redux can affect. You can debug components where used redux and see if it affects

**Comment by rubbishCoder** [comment]
@devoren  or @AndreiCalazans  if I access redux state to set the stacks and also few data to be shown in bottom tab navigator, will it affect the navigation.

**Comment by AndreiCalazans** [comment]
@rubbishCoder hard to say simply from your comment, your app is total black box to us, there could be many factors at play. 

What's not clear to me in your issue is if this was a regression that occurred after you upgraded your React Native version or if this was something you just found out that is impacting users on Samsung 23 ultra devices? 

I ask the above because there is an ongoing issue with people experience app wide performance degradation on versions 0.70.x + https://github.com/facebook/react-native/issues/36296

**Comment by rubbishCoder** [comment]
@AndreiCalazans initially I found this issue on react-native 0.71.6. so later I upgraded it to 0.72.4 it was slightly better but still bad performance. It's not only occuring in s23 ultra, it's on Samsung devices with Android version 13.

**Comment by rubbishCoder** [comment]
@AndreiCalazans @devoren when I do asynchronous api calls the response takes more time in android version 13. For example in IOS it took 1 second same call in android takes 15 secs. Any idea what will cause this.

**Comment by AndreiCalazans** [comment]
@rubbishCoder can you try profiling it and comparing the iOS profilings with the Android? That's odd if the actual network request latency is the same. The difference could be the parsing of the request response so a profiling could help indicate where the time was spent. Save the CPU profilings and share it here for us to take a look. 

[See this for](https://reactnative.dev/docs/hermes#debugging-js-on-hermes-using-google-chromes-devtools) how to use chrome://inspect to profile performance.

**Comment by cortinico** [comment]
Can we have a reproducer and a video to understand "how bad" is the regression?

**Comment by rubbishCoder** [comment]
> Can we have a reproducer and a video to understand "how bad" is the regression?

@cortinico 
Like due to security reasons I can't share the video but for sure I see a delay in the API response that happens only in latest Samsung models. Like a API that takes 500ms in iOS and other android devices, but in Samsung s23 ultra it takes around 9 seconds. It's really weird.

**Comment by devoren** [comment]
@rubbishCoder Have you debugged your API request? At what part does the query become slow?

**Comment by rubbishCoder** [comment]
> @rubbishCoder Have you debugged your API request? At what part does the query become slow?

@devoren 

I make axios call and it take 10 seconds to return a response

**Comment by cortinico** [comment]
> Like due to security reasons I can't share the video

We can't provide any support without a reproducer (when I say reproducer I don't mean **your project** but another project that behaves similarly)

**Comment by devoren** [comment]
> I make axios call and it take 10 seconds to return a response
Did you check the time from the server when calling the request? Is the request on the server delayed by 10 seconds or does the response arrive with a delay of 10 seconds?

**Comment by rubbishCoder** [comment]
> > I make axios call and it take 10 seconds to return a response
> 
> Did you check the time from the server when calling the request? Is the request on the server delayed by 10 seconds or does the response arrive with a delay of 10 seconds?

server is not delaying the response, verified it. @devoren

**Comment by rubbishCoder** [comment]
> > Like due to security reasons I can't share the video
> 
> We can't provide any support without a reproducer (when I say reproducer I don't mean **your project** but another project that behaves similarly)

will try to create a reproducer

**Comment by 4knahs** [comment]
@rubbishCoder we also have a huge increase of SIGABRT/ANR from hundreds of Samsung users, we have also reach to a point where we believe its related to August updates, just bought a new device to test. Our app crashes at random points in time, no memory leaks.

**Comment by rubbishCoder** [comment]
@cortinico @devoren  @4knahs  I have done some profiling and got some findings. I use the redux-persist and filesystem storage to persist the data in the cache directory. when the size of the cache directory is increasing the memory usage of the app is also increasing parallel. 

For example:-
Cache directory size is 120MB, Memory usage is 1GB in android devices generally.

Since I persist the data even if I kill and open or do anything the memory usage remains the same until I clear the cache directory completely. 

Let me know if you guys have any thought on it, thanks.

**Comment by cortinico** [comment]
> Let me know if you guys have any thought on it, thanks.

If you believe this is related to `redux-persist`, you should open the issue against them as here we can't really help you

**Comment by rubbishCoder** [comment]
@cortinico I tried removing persist and test in Samsung S23 its still the same behaviour no difference. Ram usage drastically got reduced but still the performance was bad in s23. So not sure which is causing the issue.

**Comment by cortinico** [comment]
Again, without a reproducer we can't help in any form

**Comment by meetp-zluck** [comment]
We are having the same issue. Most of the samsung device are lagging with 2 of our react-native apps. I've sensed the lag in from a normal device to high end devices Samsung 22 ultra. Lag is there in tabs as well. But same things are working smooth in other Brands devices from 4 years ago.

If anyone knows any workarounds or patches, I'd really appreciate. Thanks!

**Comment by PradeepKK13** [comment]
> @rubbishCoder can you try profiling it and comparing the iOS profilings with the Android? That's odd if the actual network request latency is the same. The difference could be the parsing of the request response so a profiling could help indicate where the time was spent. Save the CPU profilings and share it here for us to take a look.
> 
> [See this for](https://reactnative.dev/docs/hermes#debugging-js-on-hermes-using-google-chromes-devtools) how to use chrome://inspect to profile performance.


@cortinico @AndreiCalazans  We had done the profiling and captured the trace from Samsung S23 Ultra and Oneplus 9pro. Sharing the trace for reference. Oneplus is working fine without any lag in network or in rendering but Samsung there is lag in both network and rendering.

**Trace:** https://drive.google.com/drive/folders/1f441btiLRdcpAwtktWhBrd8QHQN2yvL9?usp=sharing

**Comment by AndreiCalazans** [comment]
@PradeepKK13 this is helpful. 

One thing that stood out while comparing those two traces is native module traces of `callJavaModuleMethod` are much slower on the Samsung S23 Ultra which is odd given that the S23 Ultra is a higher end phone. 

This seems to be due to binder transactions taking much longer.

Evidence:

Traces to `callJavaModuleMethod` are faster on Oneplus

<img width="1200" alt="Screenshot 2023-11-22 at 10 43 37" src="https://github.com/facebook/react-native/assets/20777666/39d24b15-ff3b-4858-b269-41fc4e8a0145">

*Note calls are at 36ms*



Querying for max and averages

<img width="1200" alt="Screenshot 2023-11-22 at 10 43 04" src="https://github.com/facebook/react-native/assets/20777666/b1830d1a-a668-4639-8bd1-c2b6c9b1efb2">


*Max at 113ms while average is 0.8ms*

**On S23 Ultra**

<img width="1915" alt="Screenshot 2023-11-22 at 10 44 00" src="https://github.com/facebook/react-native/assets/20777666/02647d8e-761e-4eac-9192-b1063abbc2d9">

*We see multiple `callJavaModuleMethod` calls taking 200ms and more*

<img width="1886" alt="Screenshot 2023-11-22 at 10 42 49" src="https://github.com/facebook/react-native/assets/20777666/9652f97c-f750-49dc-9ea3-0f25c12eeaad">


*Max at 300ms while average is 2.91ms*


Looking at the binder transactions we confirm slower binder transactions:

<img width="1200" alt="Screenshot 2023-11-22 at 10 48 10" src="https://github.com/facebook/react-native/assets/20777666/69e69d82-4b0e-4935-83a6-274ad70bc2e3">



@PradeepKK13 it's not obvious if this is the reason for a slower rendering and networking but it could be. Can you do the same trace on this same Samsung S23 Ultra device using the previous React Native version that had no lag? This could help us compare the two traces and validate if the binder transactions got slower or were the same.

**Comment by PradeepKK13** [comment]
> @PradeepKK13 this is helpful.
> 
> One thing that stood out while comparing those two traces is native module traces of `callJavaModuleMethod` are much slower on the Samsung S23 Ultra which is odd given that the S23 Ultra is a higher end phone.
> 
> This seems to be due to binder transactions taking much longer.
> 
> Evidence:
> 
> Traces to `callJavaModuleMethod` are faster on Oneplus
> 
> <img alt="Screenshot 2023-11-22 at 10 43 37" width="1200" src="https://user-images.githubusercontent.com/20777666/284925861-39d24b15-ff3b-4858-b269-41fc4e8a0145.png">
> _Note calls are at 36ms_
> 
> Querying for max and averages
> 
> <img alt="Screenshot 2023-11-22 at 10 43 04" width="1200" src="https://user-images.githubusercontent.com/20777666/284926547-b1830d1a-a668-4639-8bd1-c2b6c9b1efb2.png">
> _Max at 113ms while average is 0.8ms_
> 
> **On S23 Ultra**
> 
> <img alt="Screenshot 2023-11-22 at 10 44 00" width="1915" src="https://user-images.githubusercontent.com/20777666/284926696-02647d8e-761e-4eac-9192-b1063abbc2d9.png">
> _We see multiple `callJavaModuleMethod` calls taking 200ms and more_
> 
> <img alt="Screenshot 2023-11-22 at 10 42 49" width="1886" src="https://user-images.githubusercontent.com/20777666/284926894-9652f97c-f750-49dc-9ea3-0f25c12eeaad.png">
> _Max at 300ms while average is 2.91ms_
> 
> Looking at the binder transactions we confirm slower binder transactions:
> 
> <img alt="Screenshot 2023-11-22 at 10 48 10" width="1200" src="https://user-images.githubusercontent.com/20777666/284927452-69e69d82-4b0e-4935-83a6-274ad70bc2e3.png">
> Summary:
> 
> callJavaModuleMethod	OnePlus	S23 Ultra
> Average	0.8ms	2.9ms
> Max	113ms	300ms
> Observed	36ms	223ms
> @PradeepKK13 it's not obvious if this is the reason for a slower rendering and networking but it could be. Can you do the same trace on this same Samsung S23 Ultra device using the previous React Native version that had no lag? This could help us compare the two traces and validate if the binder transactions got slower or were the same.

@AndreiCalazans  Thanks for your insights. Unfortunately we don't have a React Native version in which there was no lag. Initial project was in 0.71.6 version which had the same lag and then we upgraded to 0.72.4 version which made no difference.

**Comment by diegolmello** [comment]
@PradeepKK13 can you try downgrading to 0.68?
It might help proving my point on https://github.com/facebook/react-native/issues/36296#issuecomment-1769092167 and help getting visibility to this issue.
Thanks!

**Comment by chramos** [comment]
We are using 0.64.4 and we are also experiencing a really slow performance on Samsung s23.
Hermes is turned on for a long time (2 years approx.)
honestly, I never felt any improvement on turning on/off Hermes.
I'm going to turn it off and test
I come back to let you know

**Comment by nilsrk** [comment]
Also having issues running my React Native app on a OnePlus with OxygenOS (based on Android 13.x), very laggy

**Comment by hthieu1110** [comment]
I'm facing the same issue on Android using Exynos CPU. It's working very well on an Snapdragon one.
The network call is very slow : 5s vs 100ms on iPhone but I think the source of problem is the rendering which blocks the JS thread. Do we have any news on this thread ?

**Comment by GuillaumeHemmen** [comment]
Can the stalebot be disabled on this one? This seems related to https://github.com/facebook/react-native/issues/36296 so i believe that is pertinent to keep it open.

**Comment by dorshay6** [comment]
We are also facing that issue, any idea on whats next?

**Comment by pavelustenko** [comment]
bump

**Comment by zainalipropchain** [comment]
"react-native": "0.71.2",
Android Version : 14 
Device : sumsung A52 

I am also facing same Issue , I check same screen on other devices like realme it works fine .. but same screen on sumsung is **lagging** .

**Comment by ZainaliSyed** [comment]
"react-native": "0.71.2",
Android Version : 14
Device : sumsung A52

I am also facing same Issue , I check same screen on other devices like realme it works fine .. but same screen on **sumsung is lagging .**

**Comment by AndreiCalazans** [comment]
Some have shared that version 0.73 is fixed. I suggest upgrading. I also have a prod app on 0.73 and our performance metrics have been great.

https://github.com/facebook/react-native/issues/36296#issuecomment-2103787441

**Comment by arpkirill** [comment]
> Some have shared that version 0.73 is fixed. I suggest upgrading. I also have a prod app on 0.73 and our performance metrics have been great.
> 
> 
> 
> https://github.com/facebook/react-native/issues/36296#issuecomment-2103787441

Good afternoon, have you updated the app to 0.74.2. I still have problems with Samsung in the nave

**Comment by exlymex** [comment]
Same problem with 0.74.3

**Comment by mjm918** [comment]
Any solution on this ? im using 0.74.5 and it's super slow! i upgraded from from 65 to 74 and all the users have started complaining

**Comment by AndreiCalazans** [comment]
If you still have a performance issue after updating to 0.73+ I suggest you create a separate issue with profilings and possible reproducibles that point to React Native as a culprit.

**Comment by KaitechTiendn** [comment]
Although this issue was reported some time ago, I am now facing a similar problem. Could you kindly let me know what the root cause and your solution were at that time?

I appreciate any insights you can share. Thank you very much for your time.

**Comment by AzaelContramaestre01** [comment]
Im having the samue issue

On the Samsung Galaxy S25 (Android 14), our React Native app experiences significantly slower initial mounting and fetch/network requests compared to other devices. This includes both performance degradation in the JavaScript thread and very slow response times on fetch or axios, specifically within the first 2–3 seconds of rendering.

Other Android devices (e.g., Pixel 6, Xiaomi, or emulator) perform as expected. This issue seems unique to Samsung’s new scheduling, battery/network optimization, or foreground throttling mechanisms.

We suspect a combination of:
Too many dispatches (Redux) or network requests during mount

Throttling by Samsung's scheduler

Delayed execution of JS thread during early render

We have mitigated this by splitting requests in timed setTimeout blocks or deferring them using InteractionManager.runAfterInteractions, which improved responsiveness.


https://streamable.com/mh6pev

**Comment by AzaelContramaestre01** [comment]
> Any solution on this ? im using 0.74.5 and it's super slow! i upgraded from from 65 to 74 and all the users have started complaining

How did u solve it bro?

**Comment by lakkad-amish** [comment]
i am using 0.75.0 facing the same issue. On onePlus some devices are lagging and samsung S22 and S25 Ultra is extreamly slow. anyone find solution then please let me know client is now frusted due to platform specific behaviour.

**Comment by sriharsha-rently** [comment]
Updated to React native 0.83 and tried experimental Hermes V1, still app is slow in samsung devices.

**Comment by lakkad-amish** [comment]
> Updated to React native 0.83 and tried experimental Hermes V1, still app is slow in samsung devices.

I updated to RN 0.81.5 Lagging issue fixed in samsung but in android 16 api calls taking too much. anyone facing same issue please update here. of suggest if anyone have solution

**Comment by nporchis** [comment]
same here on 0.78.3 and samsung S25 / Android 16.
but last android update of january (S931U1UES7BYLR) seems to clean the lagging issue !

**Comment by Qdafengzi** [comment]
Even after upgrading to the 0.82.1 version of the Samsung s22Ultral, it still lags abnormally. I have no idea exactly where the problem lies. It seems that this issue has been persistent in so many versions and has not been properly resolved. Sometimes, it takes 8 seconds for a button of mine to respond. Ironically, such a question has been closed

**Comment by irfan106** [comment]
We are also facing this same issue with the Samsung devices, we are using react native- 0.81.5.

**Comment by Qdafengzi** [comment]
> We are also facing this same issue with the Samsung devices, we are using react native- 0.81.5.

I have roughly identified what has affected the lag and submitted bugs in the three libraries of navigation, reanimated, and react native respectively. According to the elimination method, animation has affected the page lag. In the navigation, only by disabling the navigation can the page open and close responses. However, this is definitely not a correct solution. Additionally, code components that use animation will lag. I don't know if this is due to reanimated or react native itself. Most likely, it has nothing to do with the navigation. However, I submitted the bug just to make them take it seriously. This issue has been unresolved for a long time on the new architecture and has seriously affected google play's requirement to list the 16KB package by the end of May.

**Comment by manuelcaru** [comment]
Adding a recent data point — we're hitting this on Samsung Galaxy S24 Ultra (SM-S928B) running Android 16 / One UI 8.0 with the latest firmware (S928BXXS5CZD1, 19 April 2026, security patch 5 April 2026).

Stack: React Native 0.85.2, New Architecture enabled (Fabric + TurboModules), Hermes, all major native libraries on their latest 2026 versions (Reanimated 4.3.0, Screens 4.24.0, Gesture Handler 2.31.1, React Navigation 7.x).

Symptom: identical to what others have described — network requests are fast (~400 ms TTFB confirmed via Datadog Session Replay), JS thread is healthy (refresh rate steady at 59.98 fps, no Long Tasks captured), but queries fire in cascade with 1–3 s gaps between each. Total time from screen mount to data appearing is ~10× what the same app shows on other devices (Pixel, OnePlus, older Samsung S9 on Android 10 — all fine with the same bundle).

What we've ruled out on our side:

Firmware (user is on the latest available)
JS thread starvation (DD metrics consistently healthy)
Local cache / persistence overhead
Network latency
Our own code paths (verified by reducing query count and prefetching)
What we have not been able to mitigate:

The cascading gap between useEffect → useQuery → response → next effect. The JS thread is idle in those gaps; the time is spent at the JS↔native boundary.
This is consistent with @AndreiCalazans's earlier finding about slower callJavaModuleMethod / binder transactions on Samsung. The newer Android versions appear to have made it worse rather than better.

We've upgraded RN incrementally (0.81 → 0.85.2) over the last year and seen no meaningful change in this specific behaviour on Samsung devices. Workarounds like deferring non-critical work with requestIdleCallback, reducing query fan-out, and trimming GraphQL payloads help proportionally but don't close the gap with other OEMs.

**Comment by irfan106** [comment]
@Qdafengzi @manuelcaru 

The issue seems to be related to `react-native-keychain`. On Samsung devices, it was causing noticeable lag in our app.

After replacing `react-native-keychain` with `react-native-mmkv`, the lag issue was fixed, and the app became faster overall.

If you're using Keychain, it might be worth trying MMKV to see if it resolves the issue for you as well.

We found a few more things while debugging this issue. I'll share a detailed post about our findings later when I get some time.
